sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -47,7 +48,8 @@ case class ScalaUDF(
     inputTypes: Seq[DataType] = Nil,
     udfName: Option[String] = None,
     nullable: Boolean = true,
-    udfDeterministic: Boolean = true)
+    udfDeterministic: Boolean = true,
+    nullableTypes: Seq[Boolean] = Nil)
```

Comment:
We put the assert in the rule `HandleNullInputsForUDF`.

can we merge `inputTypes` and `nullableTypes` here so that we don't need to worry about it any more? cc @srowen

Reply:
The logic here was again that we wanted to avoid changing the binary signature. I know catalyst is effectively private to Spark, but this wasn't marked specifically private; I wondered if it would actually affect callers? If not we can go back and merge it.

Nil is just an empty list; I don't think it's dangerous and it is used above in `inputTypes`. It is not always set, because it's not always possible to infer the schema, let alone nullability.

Reply:
I think the problem is more about the way we handle `nullableTypes` if not specified as in https://github.com/apache/spark/pull/22259/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R2157. The test failure of https://github.com/apache/spark/pull/21851/files#diff-e8dddba2915a147970654aa93bee30a7R344 would have been exposed if the `nullableTypes` had been updated in this PR. So I would say logically this parameter is required, but right now it is declared optional. In this case, things went wrong when `nullableTypes` was left unspecified, and this could happen not only with tests but in "source" too. I suggest we move this parameter up right after `inputTypes` so it can get the attention it needs.

Reply:
I see, you are saying that some UDF needed to declare nullable types but didn't? I made the param optional to try to make 'migration' easier and avoid changing the signature much. But, the test you point to, doesn't it pass? are you saying it should not pass?

Reply:
Yes, the test should not pass after removing `isInstanceOf[KnownNotNull]` condition from `needsNullCheck` test (https://github.com/apache/spark/pull/22259/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2L2160). The idea was to add a `KnownNotNull` node on top of the original node to mark it as null-checked, so the rule won't add redundant null checks even if it is accidentally applied again. I'm not sure about the exact reason why you removed `isInstanceOf[KnownNotNull]` condition in this PR, but I think it should be left there alongside the new nullable type check.
After adding the `nullableTypes` parameter in the test, the issue can be reproduced:
```
  test("SPARK-24891 Fix HandleNullInputsForUDF rule") {
    val a = testRelation.output(0)
    val func = (x: Int, y: Int) => x + y
    val udf1 = ScalaUDF(func, IntegerType, a :: a :: Nil, nullableTypes = false :: false :: Nil)
    val udf2 = ScalaUDF(func, IntegerType, a :: udf1 :: Nil, nullableTypes = false :: false :: Nil)
    val plan = Project(Alias(udf2, "")() :: Nil, testRelation)
    comparePlans(plan.analyze, plan.analyze.analyze)
  }
```
BTW, I'm just curious: It looks like `nullableTypes` indicates something opposite to "nullable" used in schema. I would assume when `nullableTypes` is `Seq(false)`, it means this type is not nullable and we need not add the null check, vice versa. Did I miss something here?

Reply:
I think the logic was that `nullableTypes` has all the information about which inputs are allowed to be null. Really it's true for reference types, which can be assigned null, and false for things like primitives, which can never take the value null. It might be described as "doesntNeedNullCheck". That could be flipped to "needsNullCheck" although we'd have to go back and flip all the callers.

We can add back the check for `isInstanceOf[KnownNotNull]`; you know this area better than I.

But it doesn't quite solve the issue that anyone calling `ScalaUDF` might forget to specify which types need the null check, now that Scala 2.12 can't infer it. We can make it required by moving the args around, but I worried that would break callers that use this 'private' API? Then again, is it worse to silently still accept the ScalaUDF when its interpretation could become wrong in Scala 2.12?

Reply:
Yeah, understood and agree. We'd probably need to change 3 things in follow-up PR(s):
1. Add `isInstanceOf[KnownNotNull]` back (I can do this).
2. Move `nullableTypes` up in the parameter list and make it required. I agree that it'd only make things worse if anyone uses this private API and forget to set this parameter.
3. Flip the `true`/`false` around for `nullableTypes`. This is rather minor, but helps with readability and consistency I think.

Reply:
I'm OK with that, but obviously we need to get this into 2.4 if it's going to change. There's an RC out that's almost passed, and I flagged it. @cloud-fan it's worth considering a new RC for a change like this, even as I'm always a little uneasy about late significant changes. That said we need to release it in the way we want it. (And there are a few other minor doc things we could get into 2.4)

Reply:
Yes, I think this should be in 2.4, too, since it's an API change.
BTW, I just finished https://github.com/apache/spark/pull/22701 addressing the suggested change 2 above. Then 1 and 3 can be covered in another PR.

Reply:
what's the impact to end users?

Reply:
then why does this need to be a blocker? Private APIs keep changing very often.

Reply:
Yeah I could see an argument that it need not block release. The functionality works as intended, at least. 

Would you change it again in 2.4.1? If not then we decide to just keep this behavior. Let's say at least get this in if there is a new RC.

Reply:
Yeah I could see an argument that it need not block release. The functionality works as intended, at least. 

Would you change it again in 2.4.1? If not then we decide to just keep this behavior. Let's say at least get this in if there is a new RC.

Reply:
I think it's mostly about maintainability. We should definitely update the code as @maryannxue said, so that `ScalaUDF` is easier to use and not that error-prone. I feel we don't need to backport it, as it's basically code refactor.

Reply:
@maryannxue Please submit a PR to make the parameter `nullableTypes ` required.

Reply:
Yeah looks like we should just make these changes after all, and for 2.4, as we need another RC.